base_layer/core/src/mempool/service/service.rs

```diff
@@ -0,0 +1,375 @@
+                            ))?;
+                    let _ = reply_tx.send(Ok(response).or_else(|resp| {
+                        error!(
+                            target: LOG_TARGET,
+                            "Failed to send outbound request from Mempool service"
+                        );
+                        Err(resp)
+                    }));
+                }
+            },
+            None => {
+                info!(target: LOG_TARGET, "Discard incoming unmatched response");
+            },
+        }
+
+        Ok(())
+    }
+
+    async fn handle_outbound_request(
+        &mut self,
+        reply_tx: OneshotSender<Result<MempoolResponse, MempoolServiceError>>,
+        request: MempoolRequest,
+    ) -> Result<(), MempoolServiceError>
+    {
+        let request_key = BASE_NODE_RNG.with(|rng| generate_request_key(&mut *rng.borrow_mut()));
+        let service_request = proto::MempoolServiceRequest {
+            request_key,
+            request: Some(request.into()),
+        };
+
+        let dest_count = self
+            .outbound_message_service
+            .send_message(
+                BroadcastStrategy::Random(1),
+                NodeDestination::Unknown,
+                OutboundEncryption::EncryptForDestination,
+                OutboundDomainMessage::new(TariMessageType::MempoolRequest, service_request),
+            )
+            .await
+            .map_err(|e| MempoolServiceError::OutboundMessageService(e.to_string()))?;
+
+        if dest_count > 0 {
+            // Spawn timeout and wait for matching response to arrive
+            self.waiting_requests.insert(request_key, Some(reply_tx));
+            self.spawn_request_timeout(request_key, self.config.request_timeout)
+                .await;
+        } else {
+            let _ = reply_tx.send(Err(MempoolServiceError::NoBootstrapNodesConfigured).or_else(|resp| {
+                error!(
+                    target: LOG_TARGET,
+                    "Failed to send outbound request from Mempool service as no bootstrap nodes were configured"
+                );
+                Err(resp)
+            }));
+        }
+        Ok(())
+    }
+
+    async fn handle_incoming_transaction(
+        &mut self,
+        domain_transaction_msg: DomainMessage<Transaction>,
+    ) -> Result<(), MempoolServiceError>
+    {
+        let DomainMessage::<_> { source_peer, inner, .. } = domain_transaction_msg;
+
+        self.inbound_handlers.handle_transaction(&inner.clone().into()).await?;
+
+        self.outbound_message_service
+            .propagate(
+                NodeDestination::Unknown,
+                OutboundEncryption::None,
+                vec![source_peer.public_key],
+                OutboundDomainMessage::new(TariMessageType::NewTransaction, ProtoTransaction::from(inner)),
+            )
+            .await
+            .map_err(|e| MempoolServiceError::OutboundMessageService(e.to_string()))?;
```

Comment:
So you're suggesting that the mempool service be responsible for propagating transactions to the network rather than the base node.

I think this ok; but then the mempool should probably handle everything about transactions, including validation. Otherwise we have splitting of responsibilities.

Reply:
Yes, I was unsure about this, but it seemed like it can work. The Mempool service handles the validation and propagation of transactions and the Base Node service handles blocks.

Reply:
Great